datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -1670,17 +1734,106 @@ fn inlist_except(mut l1: InList, l2: InList) -> Result<Expr> {
     Ok(Expr::InList(l1))
 }
 
+pub struct RewriteCycle {
+    consecutive_unchanged_count: usize,
+    total_iterations: usize,
+    num_rewriters: usize,
+}
+pub type RewriteCycleResult = ControlFlow<Result<Expr>, Expr>;
+
+impl RewriteCycle {
+    fn new() -> Self {
+        RewriteCycle {
+            // use usize::MAX as default to avoid checking null in is_done() comparison
+            // real value is set later by record_num_rewriters
+            num_rewriters: usize::MAX,
+            consecutive_unchanged_count: 0,
+            total_iterations: 0,
+        }
+    }
+
+    pub fn completed_cycles(&self) -> usize {
+        // default value indicates we have not completed a cycle
+        if self.num_rewriters == usize::MAX {
+            0
+        } else {
+            self.total_iterations / self.num_rewriters
+        }
+    }
+
+    pub fn total_iterations(&self) -> usize {
+        self.total_iterations
+    }
+
+    fn record_num_rewriters(&mut self) {
+        self.num_rewriters = self.total_iterations;
+    }
+
+    fn is_done(&self) -> bool {
+        self.consecutive_unchanged_count >= self.num_rewriters
```

Comment:
The logic to stop looping is checking if `transformed` meet the num of rewriters, but can we just stop if any of the `transformed` is false? I think it is more straightforward and we don't need to have a `max_iter` num too.

Idea is something like
```rust
loop {
  transformed |= rewrite_rule()
  if not transformed {
     return
  }
}
```

Reply:
🤔 I think there are cases when simplification will do something that allows more constant propagation to proceed so even if `transformed=false` is returned another application of a different pass could actually simplify things -- so in this case for example if the const evaluator returns false, simpliy could still return true

Reply:
> if the const evaluator returns false, simpliy could still return true

In this case, `transformed` is true, and we should try another pass.

I think once we found `transformed` is false for every rules, any further pass will still remain unchanged, so we can skip it.

Reply:
> so in this case for example if the const evaluator returns false, simpliy could still return true

In this case, transformed is true, so we should run another pass. 
I think once we found every rules is not transformed, we can stop the passes

Reply:
> I think once we found every rules is not transformed in a pass that means we done the optimization

Yes, I agree -- I think this is what the counter is attempting to do

Reply:
> 🤔 I think there are cases when simplification will do something that allows more constant propagation to proceed so even if `transformed=false` is returned another application of a different pass could actually simplify things -- so in this case for example if the const evaluator returns false, simpliy could still return true

yes I found this to be true when running tests against this code. an additional constant evaluation produced a new transformation